Sage/carbon
# feat(select): add virtual scrolling support - FE-5394

### Proposed behaviour



Add support for "virtual scrolling" on Select and its variants.

This PR adds 2 new props to all 3 Select components:
- `enableVirtualScroll`: a boolean prop that when set enables virtual scrolling - meaning not all children of the SelectList will be rendered in the DOM at any one time, but instead only those currently visible on the page, plus a few either side
- `virtualScrollOverscan` is the number of children to be rendered into the DOM either side of the visible ones (this defaults to 5). Has no effect when `enableVirtualScroll` is not used.

Also fix some accessibility issues:
- remove some Axe errors for the `multiColumn` mode on Select
- ensure that the guid IDs of select options in the DOM remains stable under any prop change of Select, and only change in the rare case where children are added or removed dynamically. In particular, in a controlled Select, selecting an option does not change the IDs of the children, and so the Axe error about an invalid value for aria-activedescendant is no longer present.

### Current behaviour

Selects always render all their children into the DOM, leading to severe performance issues when there are thousands of children.

There are several ARIA-related errors when Axe is run with the Select List open in a multicolumn select. Also in a controlled select the IDs of children change when selecting an option, leading to an Axe error about ARIA-activedescendant.



### Checklist




#### QA


### Additional context



Note for UX QA - there is a small difference in behaviour after this PR on multiColumn selects. Previously only the table body was scrollable, so the scrollbar does not go up to the sticky table head. In order to make scrolling work correctly in this PR I had to make the whole table scrollable so the scrollbar extends to the table header, even though that stays fixed as you scroll. I don't believe this is an issue and don't think I can avoid it, but please let me know if you think it's a serious problem.

### Testing instructions



Test with the new `virtualScrollBuffer` prop on Selects with large numbers of options (the new "virtualised" story is a simple example), and ensure that it scrolls smoothly and performance is good. (Compare that to performance without the prop to understand why this was needed!)

Perform full regression testing on all 3 types of Select (Simple, Filterable, Multi). Test all stories and all props, including with keyboard navigation through the list.

Also test with screenreaders, in all supported screenreader/browser combinations, to ensure that when the new prop is set the component is still usable and there are no major accessibility issues (or at least not ones that didn't exist before!).

NOTE: when running Axe with the Select List open there is an error "scrollable region must have keyboard access", which did not exist before (except on a multi-column select, where it did). This refers to the select list itself and is a false positive (I checked this with @tempertemper) because it's possible to navigate to the list with the keyboard using the select textbox, as before. I've tried to alter the markup to remove this error but it isn't possible to do this while still getting the desired behaviour. I believe this is OK as it isn't an actual usability barrier and is also only rated "moderate" by Axe, but we should note that the error exists.
